### PR TITLE
Add template functions for Bcrypt and Argon2

### LINKED ISSUE #1689: Templates should support modern password hashing functions
We should consider adding support for more modern password hashing functions like Argon2 or Bcrypt, too.

https://godoc.org/golang.org/x/crypto/argon2
https://godoc.org/golang.org/x/crypto/bcrypt
https://doc.dovecot.org/configuration_manual/authentication/password_schemes/

### PR DESCRIPTION
**Pull request by dominikschulz**
Fixes #1689

RELEASE_NOTES=[ENHANCEMENT] Add tpl funcs for Bcrypt and Argon2

Signed-off-by: Dominik Schulz <dominik.schulz@gauner.org>

### CODE REVIEW

**Comment by dominikschulz** [comment]
This PR doesn't come with tests as I deliberately didn't want to implement validation for these schemes, but I've did end to end tests with Dovecot and that works.

File: internal/pwschemes/argon2i/argon2i.go [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+	"crypto/rand"
+	"encoding/base64"
+	"fmt"
+
+	"golang.org/x/crypto/argon2"
+)
+
+const (
+	time    uint32 = 4
+	memory  uint32 = 32 * 1024
```
Comment (AnomalRoil) [inline internal/pwschemes/argon2i/argon2i.go:13]: As per section 8 and 9 of the [Argon2 paper](https://github.com/P-H-C/phc-winner-argon2/blob/master/argon2-specs.pdf), I think we might want to increase this, unless it's the settings used by Dovecot?

File: internal/pwschemes/argon2id/argon2id.go [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+	"crypto/rand"
+	"encoding/base64"
+	"fmt"
+
+	"golang.org/x/crypto/argon2"
+)
+
+const (
+	time    uint32 = 3
+	memory  uint32 = 64 * 1024
```
Comment (AnomalRoil) [inline internal/pwschemes/argon2id/argon2id.go:13]: See previous comment. 64MB is low compared to recommended values.

File: internal/tpl/funcs.go [hunk 3]
```diff
@@ -27,6 +30,9 @@ const (
 	FuncGet         = "get"
 	FuncGetPassword = "getpw"
 	FuncGetValue    = "getval"
+	FuncArgon2i     = "argon2i"
```
Comment (AnomalRoil) [inline internal/tpl/funcs.go:33]: Please document these new functions in the Template.md documentation.

**Review by AnomalRoil — CHANGES_REQUESTED** [review]

File: internal/pwschemes/argon2i/argon2i.go (hunk 1 above)
Comment (dominikschulz) [reply internal/pwschemes/argon2i/argon2i.go:13]: No, these values are encoded into the hash and Dovecot will re-use whatever we use here. I haven't read the whole paper so I'm not sure if I understand those correctly, yet. But I'll start with a little more and we can change it or make it configurable later.

File: internal/pwschemes/argon2id/argon2id.go (hunk 2 above)
Comment (dominikschulz) [reply internal/pwschemes/argon2id/argon2id.go:13]: Done

File: internal/tpl/funcs.go (hunk 3 above)
Comment (dominikschulz) [reply internal/tpl/funcs.go:33]: Done.

**Pull request merged**